File: libgomp/testsuite/libgomp.c/examples-4/async_target-2.c
/* { dg-do run } */
/* { dg-require-effective-target offload_device_nonshared_as } */

#include <omp.h>
#include <stdlib.h>

#define EPS 0.00001
#define N 10000

#pragma omp declare target
void init (float *a, float *b, int n)
{
  int i;
  for (i = 0; i < n; i++)
    {
      a[i] = 0.1 * i;
      b[i] = 0.01 * i * i;
    }
}
#pragma omp end declare target

void vec_mult_ref(float *p, int n)
{
  float *v1, *v2;
  int i;

  v1 = (float *) malloc (n * sizeof (float));
  v2 = (float *) malloc (n * sizeof (float));

  init (v1, v2, n);

  for (i = 0; i < n; i++)
    p[i] = v1[i] * v2[i];

  free (v1);
  free (v2);
}

void vec_mult(float *p, int n)
{
  float *v1, *v2;
  int i;

  #pragma omp task shared(v1, v2) depend(out: v1, v2)
    #pragma omp target map(v1, v2)
      {
	if (omp_is_initial_device ())
	  abort ();

	v1 = (float *) malloc (n * sizeof (float));
	v2 = (float *) malloc (n * sizeof (float));

	init (v1, v2, n);
      }

  #pragma omp task shared(v1, v2) depend(in: v1, v2)
    #pragma omp target map(to: v1, v2) map(from: p[0:n])
      {
	if (omp_is_initial_device ())
	  abort ();

	#pragma omp parallel for
	  for (i = 0; i < n; i++)
	    p[i] = v1[i] * v2[i];

	  free (v1);
	  free (v2);
      }

  #pragma omp taskwait
}

void check (float *a, float *b, int n)
{
  int i;
  for (i = 0 ; i < n ; i++)
    {
      float err = (a[i] == 0.0) ? b[i] : (b[i] - a[i]) / a[i];
      if (((err > 0) ? err : -err) > EPS)
	abort ();
    }
}

int main ()
{
  float *p1 = (float *) malloc (N * sizeof (float));
  float *p2 = (float *) malloc (N * sizeof (float));

  vec_mult_ref (p1, N);
  vec_mult (p2, N);

  check (p1, p2, N);

  free (p1);
  free (p2);

  return 0;
}

File: libgomp/testsuite/libgomp.fortran/examples-4/async_target-2.f90
! { dg-do run }
! { dg-require-effective-target offload_device_nonshared_as }

subroutine init (v1, v2, N)
  !$omp declare target
  integer :: i, N
  real :: v1(N), v2(N)
  do i = 1, N
    v1(i) = i + 2.0
    v2(i) = i - 3.0
  end do
end subroutine

subroutine check (p, N)
  integer :: i, N
  real, parameter :: EPS = 0.00001
  real :: diff, p(N)
  do i = 1, N
    diff = p(i) - (i + 2.0) * (i - 3.0)
    if (diff > EPS .or. -diff > EPS) STOP 1
  end do
end subroutine

subroutine vec_mult (p, N)
  use omp_lib, only: omp_is_initial_device
  real :: p(N)
  real, allocatable :: v1(:), v2(:)
  integer :: i
  !$omp declare target (init)
  !$omp target data map(to: v1, v2, N) map(from: p)
    !$omp task shared(v1, v2, p) depend(out: v1, v2)
      !$omp target map(to: v1, v2, N)
        if (omp_is_initial_device ()) STOP 2
        allocate (v1(N), v2(N))
        call init (v1, v2, N)
      !$omp end target
    !$omp end task
    !$omp task shared(v1, v2, p) depend(in: v1, v2)
      !$omp target map(to: v1, v2, N) map(from: p)
        if (omp_is_initial_device ()) STOP 3
        !$omp parallel do
        do i = 1, N
          p(i) = v1(i) * v2(i)
        end do
        deallocate (v1, v2)
      !$omp end target
    !$omp end task
  !$omp end target data

  !$omp taskwait
  call check (p, N)
end subroutine

program e_55_2
  integer, parameter :: N = 1000
  real :: p(N)
  call vec_mult (p, N)
end program
